### PR TITLE
Detect borrow error involving sub-slices and suggest `split_at_mut`

### PR DESCRIPTION
**Pull request by estebank**
```
error[E0499]: cannot borrow `foo` as mutable more than once at a time
  --> $DIR/suggest-split-at-mut.rs:13:18
   |
LL |     let a = &mut foo[..2];
   |                  --- first mutable borrow occurs here
LL |     let b = &mut foo[2..];
   |                  ^^^ second mutable borrow occurs here
LL |     a[0] = 5;
   |     ---- first borrow later used here
   |
   = help: use `.split_at_mut(position)` or similar method to obtain two mutable non-overlapping sub-slices
```

Address most of #58792.

For follow up work, we should emit a structured suggestion for cases where we can identify the exact `let (a, b) = foo.split_at_mut(2);` call that is needed.

### CODE REVIEW

File: tests/ui/borrowck/borrowck-overloaded-index-autoderef.stderr [hunk 1]
```diff
@@ -17,6 +17,8 @@ LL |     let q = &mut f[&s];
    |                  ^ second mutable borrow occurs here
 LL |     p.use_mut();
    |     - first borrow later used here
+   |
+   = help: use `.split_at_mut(position)` or similar method to obtain two mutable non-overlapping sub-slices
```
Comment (fee1-dead) [inline tests/ui/borrowck/borrowck-overloaded-index-autoderef.stderr:21]: This seem to be a regression. We should check that the indexing is actually done on a slice and not a user custom type.

**Review by fee1-dead — CHANGES_REQUESTED** [review]

File: tests/ui/borrowck/borrowck-overloaded-index-autoderef.stderr (hunk 1 above)
Comment (estebank) [reply tests/ui/borrowck/borrowck-overloaded-index-autoderef.stderr:21]: I can easily check whether we have a `[]` or `Vec` and be more accurate that `split_at_mut` will work. We can also look for all methods on any custom type that being indexed on that returns `(&mut T, &mut T)` or `(&mut T, T)`. I was hoping not doing that as part of this PR, but I can.

If the comment instead means "they are doing `p.use_mut()`, where's the indexing, it happens earlier.

I am of the opinion that any type that supports indexing should have a method akin to `split_at_mut`. We should likely customize the error for types from the local crate, as *those* are much less likely to have an analogue.

File: compiler/rustc_hir/src/hir.rs [hunk 2]
```diff
@@ -1811,6 +1811,41 @@ impl Expr<'_> {
         }
     }
 
+    /// Whether this and the `other` expression are the same for purposes of an indexing operation.
+    ///
+    /// This is only used for diagnostics to see if we have things like `foo[i]` where `foo` is
+    /// borrowed multiple times with `i`.
+    pub fn equals(&self, other: &Expr<'_>) -> bool {
```
Comment (fee1-dead) [inline compiler/rustc_hir/src/hir.rs:1818]: The naming seems confusing as it doesn't specify that this is only in terms of indexing operations. Perhaps rename to `equivalent_for_indexing`?

**Review by fee1-dead — CHANGES_REQUESTED** [review]
One small nit, otherwise looks good to me, thanks! r=me.

**Pull request merged**